File: go/internal/eventtime/eventtime.go
// Package eventtime is the one place that knows what timezone this event's clocks are in.
//
// # Why this exists (task 358)
//
// The maintainer's instruction, 2026-09-21: *"all timestamps shall be printed in CEST tz"*. Before it, every
// timestamp on the public site was rendered in **UTC** — two hours early all through the night the pages are
// about. Measured rather than reasoned: patrol 71's first scan is epoch 1789761524, which is 21:58:44 CEST, and
// the patrol page printed *kl. 19:58*.
//
// The cause is ordinary and worth naming, because it is how this comes back: nothing was *converting* anything.
// Scans are stored as Unix seconds, `time.Unix` returns an instant in the process's local zone, the container has
// no `/etc/timezone`, so local is UTC — and the formatter printed whatever zone its input happened to carry. A
// rendering that inherits its timezone from its input has no timezone at all.
//
// # CEST, or Europe/Copenhagen?
//
// The instruction says CEST. This loads **Europe/Copenhagen**, which renders CEST in summer and CET in winter,
// and that is the instruction rather than a deviation from it: CEST *is* Copenhagen's clock in September, when
// the event happens. Hardcoding +02:00 would be wrong for a page still being read in November — and the page is
// the archive, so it will be.
package eventtime

import (
	"fmt"
	"sync"
	"time"
)

// zone is the IANA name. One constant, so a grep for the zone finds one answer.
const zone = "Europe/Copenhagen"

var (
	once     sync.Once
	location *time.Location
	loadErr  error
)

// Location is the timezone every rendered timestamp is converted to.
//
// Loaded once and cached: `LoadLocation` reads the zoneinfo database from disk, and this is called for every
// timestamp on every public page.
//
// # The fallback is UTC, and it is a deployment fault
//
// A missing zoneinfo database (a scratch image without tzdata) leaves this returning UTC, which is the wrong
// clock by an hour or two — but it is the same wrong clock everywhere, and every alternative is worse: panicking
// takes the site down over a cosmetic fault, and a hardcoded offset would be silently wrong twice a year
// forever. LoadErr exists so a caller can say so out loud at boot rather than leaving it to be found on a page.
func Location() *time.Location {
	once.Do(func() {
		loc, err := time.LoadLocation(zone)
		if err != nil {
			location, loadErr = time.UTC, err
			return
		}
		location = loc
	})
	return location
}

// LoadErr reports why Location fell back to UTC, or nil.
//
// Checked at boot so the failure is a log line somebody can act on. Nothing renders differently because of it.
func LoadErr() error {
	Location()
	return loadErr
}

// In converts an instant to the event's zone.
//
// The whole point of the package in one line: **call this at every rendering boundary**, not where a value is
// read from the database or the stream. Instants stay instants internally; a timezone is a property of how a
// human is shown a moment, not of the moment.
func In(t time.Time) time.Time { return t.In(Location()) }

// danishMonths are the month names, lowercase as Danish writes them.
//
// Not `time.Format`, and this is the second bug this package fixed: `glimtpublic.go` formatted dates with the
// layout `"2. januar 2006 kl. 15:04"`, and **`januar` is not a layout token**. Go copied it through as a
// literal, so every public glimt was dated in January — a September photograph read *"18. januar 2026"*. It
// carried a comment claiming it matched the other surface's format, which is exactly what stopped anybody
// checking.
//
// Go's own month names are English and it has no locale support, so a table is the only honest option. A test
// walks all twelve.
var danishMonths = [12]string{
	"januar", "februar", "marts", "april", "maj", "juni",
	"juli", "august", "september", "oktober", "november", "december",
}

// Danish renders an instant as the public pages say it: "18. september 2026 kl. 21:58".
//
// Converts first, always. A 24-hour clock, because that is what a Dane reads and because an event running past
// midnight makes am/pm a puzzle.
func Danish(t time.Time) string {
	t = In(t)
	return fmt.Sprintf("%d. %s %d kl. %02d:%02d",
		t.Day(), danishMonths[int(t.Month())-1], t.Year(), t.Hour(), t.Minute())
}

// Clock renders just the time of day: "21:58".
//
// For a list whose heading already establishes the date. Converts, like everything here.
func Clock(t time.Time) string {
	t = In(t)
	return fmt.Sprintf("%02d:%02d", t.Hour(), t.Minute())
}

File: vue/src/helpers/eventTime.ts
/**
 * The event's timezone, and every clock the app prints (task 358).
 *
 * # Why this module exists
 *
 * The maintainer's instruction, 2026-09-21: *"all timestamps shall be printed in CEST tz"*. The BFF had a real
 * bug — the public pages rendered UTC, two hours early — and the app had a subtler one: every formatter here
 * asked `Intl` for `da-DK` and **no timezone**, which means the *device's*. That is right for a phone in Denmark
 * and wrong for every other case, and the cases are not exotic:
 *
 *   - A phone whose timezone is wrong, or still set to somewhere a family drove back from.
 *   - A leader abroad reading a patrol's scan list, who would see a different night from the patrol's.
 *   - A screenshot in a bug report, which is now unambiguous about what clock produced it.
 *
 * The event happens on one clock. Two people looking at the same scan list should read the same time, and it
 * should be the time on the wall at the checkpoint.
 *
 * # CEST versus Europe/Copenhagen
 *
 * The zone, not the offset — matching `internal/eventtime` on the server, for the same reason: `Europe/Copenhagen`
 * is CEST in September and CET in winter, and pinning +02:00 would go wrong in late October while the pages are
 * still being read.
 */

/** The IANA zone. One constant so a grep finds one answer, and so the spec can assert on it. */
export const EVENT_TIME_ZONE = 'Europe/Copenhagen'

/** What every formatter here starts from. */
const base: Intl.DateTimeFormatOptions = { timeZone: EVENT_TIME_ZONE }

/**
 * Builds a formatter pinned to the event's timezone.
 *
 * Exported so a caller with an unusual shape does not have to reach for `Intl` directly and forget the zone —
 * which is the mistake this module exists to make hard. `eventTime.spec.ts` fails the build if any source file
 * constructs a date formatter without it.
 */
export function eventFormat(options: Intl.DateTimeFormatOptions): Intl.DateTimeFormat {
  return new Intl.DateTimeFormat('da-DK', { ...base, ...options })
}

const clockFormat = eventFormat({ hour: '2-digit', minute: '2-digit' })
const clockSecondsFormat = eventFormat({ hour: '2-digit', minute: '2-digit', second: '2-digit' })
const weekdayClockFormat = eventFormat({ weekday: 'short', hour: '2-digit', minute: '2-digit' })
const dayMonthFormat = eventFormat({ day: 'numeric', month: 'short' })

/** A time of day: `21:58`. */
export function clock(at: Date | number): string {
  return clockFormat.format(at)
}

/** A time of day with seconds: `21:58:44`. For diagnostics, where the seconds are the point. */
export function clockWithSeconds(at: Date | number): string {
  return clockSecondsFormat.format(at)
}

/** A weekday and a time: `fre. 21.58`. For a night that crosses midnight, where the day matters. */
export function weekdayClock(at: Date | number): string {
  return weekdayClockFormat.format(at)
}

/** A day and a month: `18. sep.`. For something older than a week, where the clock no longer helps. */
export function dayMonth(at: Date | number): string {
  return dayMonthFormat.format(at)
}
